File: backend/src/main/java/eu/zkkn/android/kaktus/backend/FcmSender.java
package eu.zkkn.android.kaktus.backend;

import com.google.android.gcm.server.Constants;
import com.google.android.gcm.server.Message;
import com.google.android.gcm.server.Result;
import com.google.android.gcm.server.Sender;
import com.google.appengine.api.taskqueue.QueueFactory;
import com.google.appengine.api.taskqueue.RetryOptions;
import com.google.appengine.api.taskqueue.TaskOptions;

import java.io.IOException;
import java.util.logging.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;


/**
 * Firebase cloud messages sender
 * If used as a Push Queue Task the limit for execution is 10 min, otherwise it's 1 min
 */
public class FcmSender extends HttpServlet {

    private static final Logger LOG = Logger.getLogger(FcmSender.class.getName());

    /**
     * Api Keys can be obtained from the google cloud console
     */
    private static final String API_KEY = System.getProperty("gcm.api.key");

    private static final String PARAM_MESSAGE_NAME = "msg";


    public static void sendFcmToAll(String message) {
        QueueFactory.getDefaultQueue().add(
                TaskOptions.Builder.withUrl("/tasks/fcm-sender")
                        .param(PARAM_MESSAGE_NAME, message)
                        .retryOptions(RetryOptions.Builder.withTaskRetryLimit(3)));
    }


    @Override
    protected void doPost(HttpServletRequest req, HttpServletResponse resp) {

        LOG.info("Start sending FCMs");

        @Nullable
        String message = req.getParameter(PARAM_MESSAGE_NAME);

        if (message != null && message.trim().length() != 0) {
            sendMessage(message);
        } else {
            LOG.warning("The message to send it is empty.");
        }

        LOG.info("Finish Sending FCMs");

    }


    /**
     * Send FCM to all registered devices
     *
     * @param message The message to send
     */
    private void sendMessage(@Nonnull String message) {

        //TODO: Migrate from legacy HTTP to HTTP v1 API
        // So the analytics labels could be added to messages in order to have data for reports
        // in Firebase console
        // https://firebase.google.com/docs/cloud-messaging/understand-delivery#adding_analytics_labels_to_messages
        Sender sender = new Sender(API_KEY);

        Message msg = new Message.Builder()
                .addData("type", "notification")
                .addData("message", Utils.cropText(message, 1000))
                .addData("uri", CheckServlet.KAKTUS_DOBIJECKA_URL)
                .priority(Message.Priority.HIGH)
                //.dryRun(true)
                .build();

        // Send message to topic for notifications
        String topicName = Utils.isProduction() ? "notifications" : "notifications-debug";
        String topicFullName = Constants.TOPIC_PREFIX + topicName;
        LOG.info("Send message to topic: " + topicFullName);
        Result topicResult = trySendMessage(sender, msg, topicFullName);
        if (topicResult == null || topicResult.getMessageId() == null) {
            LOG.severe("Error when sending message to " + topicFullName);
        }
        if (topicResult != null) LOG.info(topicResult.toString());

    }

    @Nullable
    private Result trySendMessage(Sender sender, Message msg, String registrationId) {
        Result result = null;
        try {
            result = sender.send(msg, registrationId, 5);
        } catch (IOException e) {
            LOG.warning(String.format("Message couldn't be sent: %s %s",
                    e.getClass().getName(), e.getMessage()));
        }
        return result;
    }

}

File: backend/src/main/java/eu/zkkn/android/kaktus/backend/FcmSender.kt
package eu.zkkn.android.kaktus.backend

import com.google.appengine.api.taskqueue.QueueFactory
import com.google.appengine.api.taskqueue.RetryOptions
import com.google.appengine.api.taskqueue.TaskOptions
import com.google.auth.oauth2.GoogleCredentials
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.messaging.AndroidConfig
import com.google.firebase.messaging.FcmOptions
import com.google.firebase.messaging.FirebaseMessaging
import com.google.firebase.messaging.Message
import java.util.logging.Logger
import javax.servlet.http.HttpServlet
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse


/**
 * Firebase cloud messages sender
 * If used as a Push Queue Task the limit for execution is 10 min, otherwise it's 1 min
 */
class FcmSender : HttpServlet() {

    private val log = Logger.getLogger(this::class.java.name)

    private val firebaseMessaging: FirebaseMessaging by lazy {
        val googleCredentials = GoogleCredentials.fromStream(
            ServletContextHolder.getServletContext()
                .getResourceAsStream("/WEB-INF/serviceAccountKey.json")
        )
        val options = FirebaseOptions.builder().setCredentials(googleCredentials).build()
        FirebaseMessaging.getInstance(FirebaseApp.initializeApp(options))
    }

    override fun doPost(req: HttpServletRequest, resp: HttpServletResponse) {
        log.info("Start sending FCMs")
        val message = req.getParameter(PARAM_MESSAGE_NAME)
        if (message != null && message.trim().isNotEmpty()) {
            sendMessage(message)
        } else {
            log.warning("The message to send is empty.")
        }
        log.info("Finish Sending FCMs")
    }

    /**
     * Send FCM to all registered devices
     *
     * @param message The message to send
     */
    private fun sendMessage(message: String) {
        // Send message to topic for notifications
        val topicName = if (Utils.isProduction()) "notifications" else "notifications-debug"
        val fcmMessage = Message.builder()
            .setTopic(topicName)
            .putData("type", "notification")
            .putData("message", Utils.cropText(message, 1000))
            .putData("uri", CheckServlet.KAKTUS_DOBIJECKA_URL)
            .setAndroidConfig(
                AndroidConfig.builder()
                    .setPriority(AndroidConfig.Priority.HIGH)
                    .build()
            )
            .setFcmOptions(FcmOptions.withAnalyticsLabel(topicName))
            .build()


        log.info("Send message to topic: $topicName")
        val messageId = send(fcmMessage)
        if (messageId.isNullOrEmpty()) {
            log.severe("Error when sending message to $topicName")
        }
        if (messageId != null) log.info("Message ID: $messageId")
    }

    private fun send(message: Message): String? {
        // perform only a dry run if not in production
        val dryRun = !Utils.isProduction()
        if (dryRun) log.warning("FCM messages are sent only from Production environment")
        return firebaseMessaging.send(message, dryRun)
    }


    companion object {

        private const val PARAM_MESSAGE_NAME = "msg"

        @JvmStatic
        fun sendFcmToAll(message: String?) {
            QueueFactory.getDefaultQueue().add(
                TaskOptions.Builder.withUrl("/tasks/fcm-sender")
                    .param(PARAM_MESSAGE_NAME, message)
                    .retryOptions(RetryOptions.Builder.withTaskRetryLimit(3))
            )
        }

    }

}
